src/oslogin_utils.cc

```diff
@@ -813,80 +821,83 @@ bool FindGroup(struct group* result, BufferManager* buf, int* errnop) {
       pageToken = "";
     }
 
-    groups.clear();
-    if (!ParseJsonToGroups(response, &groups) || groups.empty()) {
+    if (!ParseJsonToGroups(response, groups)) {
       *errnop = ENOENT;
       return false;
     }
-
-    // Check for a match.
-    for (int i = 0; i < (int) groups.size(); i++) {
-      Group el = groups[i];
-      if ((result->gr_name != NULL) && (string(result->gr_name) == el.name)) {
-        // Set the name even though it matches because the final string must
-        // be stored in the provided buffer.
-        if (!buf->AppendString(el.name, &result->gr_name, errnop)) {
-          return false;
-        }
-        result->gr_gid = el.gid;
-        return true;
-      }
-      if ((result->gr_gid != 0) && (result->gr_gid == el.gid)) {
-        if (!buf->AppendString(el.name, &result->gr_name, errnop)) {
-          return false;
-        }
-        return true;
-      }
-    }
-  } while (pageToken != "0");
-  // Not found.
-  *errnop = ENOENT;
-  return false;
+  } while (pageToken != "");
+  return true;
 }
 
-bool GetGroupsForUser(string username, std::vector<Group>* groups, int* errnop) {
+bool GetGroupByName(string name, struct group* result, BufferManager* buf, int* errnop) {
+  std::stringstream url;
+  std::vector<Group> groups;
+
   string response;
-  if (!(GetUser(username, &response))) {
-    DEBUG("GetGroupsForUser: !GetUser\n");
-    *errnop = ENOENT;
+  long http_code;
+  string pageToken = "";
+
+  url.str("");
+  url << kMetadataServerUrl << "groups?groupname=" << name;
+
+  response.clear();
+  http_code = 0;
+  if (!HttpGet(url.str(), &response, &http_code) || http_code != 200 ||
+      response.empty()) {
+    *errnop = EAGAIN;
     return false;
   }
 
-  string email;
-  if (!ParseJsonToEmail(response, &email) || email.empty()) {
-    DEBUG("GetGroupsForUser: !ParseJsonToEmail\n");
+  groups.clear();
+  if (!ParseJsonToGroups(response, &groups) || groups.empty() || groups.size() != 1) {
     *errnop = ENOENT;
     return false;
   }
 
+  Group el = groups[0];
+  result->gr_gid = el.gid;
+  if (!buf->AppendString(el.name, &result->gr_name, errnop)) {
+    return false;
+  }
+
+  return true;
+}
+
+bool GetGroupByGID(int gid, struct group* result, BufferManager* buf, int* errnop) {
   std::stringstream url;
+  std::vector<Group> groups;
 
+  string response;
   long http_code;
   string pageToken = "";
 
-  do {
-    url.str("");
-    url << kMetadataServerUrl << "groups?email=" << email;
-    if (pageToken != "")
-      url << "&pagetoken=" << pageToken;
+  url.str("");
+  char gidstring[65];
```

Comment:
Why is 65 hardcoded?

Reply:
it is the maximum string length of a 64-bit integer.